pkg/ingester/errors.go

```diff
@@ -78,111 +70,230 @@ func (e errorWithStatus) GRPCStatus() *status.Status {
+	return perUserLimitReachedError{
+		limit: limit,
+		kind:  seriesKind,
+	}
+}
+
+// newPerUserMetadataLimitReachedError creates a new perUserLimitReachedError indicating that a per-user metadata limit has been reached.
+func newPerUserMetadataLimitReachedError(limit int) perUserLimitReachedError {
+	return perUserLimitReachedError{
+		limit: limit,
+		kind:  metadataKind,
+	}
+}
+
+func (e perUserLimitReachedError) Error() string {
+	switch e.kind {
+	case seriesKind:
+		return globalerror.MaxSeriesPerUser.MessageWithPerTenantLimitConfig(
+			fmt.Sprintf("per-user series limit of %d exceeded", e.limit),
+			validation.MaxSeriesPerUserFlag,
+		)
+	case metadataKind:
+		return globalerror.MaxMetadataPerUser.MessageWithPerTenantLimitConfig(
+			fmt.Sprintf("per-user metric metadata limit of %d exceeded", e.limit),
 			validation.MaxMetadataPerUserFlag,
-		),
-	)
+		)
+	default:
+		return fmt.Sprintf("per-user limit unknown error kind: %d", e.kind)
+	}
+}
+
+// perMetricLimitReachedError is an error indicating that a per-metric limit has been reached.
+type perMetricLimitReachedError struct {
+	limit  int
+	series string
+	kind   limitKind
+}
+
+// newPerMetricSeriesLimitReachedError creates a new perMetricLimitReachedError indicating that a per-metric series limit has been reached.
+func newPerMetricSeriesLimitReachedError(limit int, labels labels.Labels) perMetricLimitReachedError {
+	return perMetricLimitReachedError{
+		limit:  limit,
+		series: labels.String(),
+		kind:   seriesKind,
+	}
 }
 
-func formatMaxMetadataPerMetricError(limits *validation.Overrides, labels labels.Labels, userID string) error {
-	globalLimit := limits.MaxGlobalMetadataPerMetric(userID)
-	return safeToWrapError(
-		fmt.Sprintf("%s This is for series %s",
+// newPerMetricMetadataLimitReachedError creates a new perMetricLimitReachedError indicating that a per-metric metadata limit has been reached.
+func newPerMetricMetadataLimitReachedError(limit int, labels labels.Labels) perMetricLimitReachedError {
+	return perMetricLimitReachedError{
+		limit:  limit,
+		series: labels.String(),
+		kind:   metadataKind,
+	}
+}
+
+func (e perMetricLimitReachedError) Error() string {
+	switch e.kind {
+	case seriesKind:
+		return fmt.Sprintf("%s This is for series %s",
+			globalerror.MaxSeriesPerMetric.MessageWithPerTenantLimitConfig(
+				fmt.Sprintf("per-metric series limit of %d exceeded", e.limit),
+				validation.MaxSeriesPerMetricFlag,
+			),
+			e.series,
+		)
+	case metadataKind:
+		return fmt.Sprintf("%s This is for series %s",
 			globalerror.MaxMetadataPerMetric.MessageWithPerTenantLimitConfig(
-				fmt.Sprintf("per-metric metadata limit of %d exceeded", globalLimit),
+				fmt.Sprintf("per-metric metadata limit of %d exceeded", e.limit),
 				validation.MaxMetadataPerMetricFlag,
 			),
-			labels.String(),
-		),
-	)
+			e.series,
+		)
+	default:
+		return fmt.Sprintf("per-metric limit unknown error kind: %d", e.kind)
+	}
+}
```

Comment:
I don't think that these errors with `kind` add more simplicity or clarity to the code, and they're also not shorter in the implementation. I'd remove that and have one error type for each error.